add user guide

eguiraud opened the pull request:
First version of the user guide. Submitting it as a PR for easier review and discussion.

The "raw" view of the file on Github looks somewhat fancier and features a more complete table of contents.

eguiraud committed:
add user guide

eguiraud committed:
fix typo and markdown linter warnings

eguiraud committed:
fix ForeachSlot example

related note: std::accumulate deduces the type from the **init** parameter,
**not** from the type of the elements of the collection. ugh.

eguiraud committed:
ForeachSlot is not coming soon anymore

bluehood commented:
![](https://github.trello.services/images/mini-trello-icon.png) [finalise user guide](https://trello.com/c/AWCxdhL2/1-finalise-user-guide)

eguiraud committed:
add introduction section

eguiraud committed:
add `Sum` action to action overview

bluehood committed:
improve introduction

dpiparo committed:
Improve the already excellent text

Some examples of improvements:
- comment out the parts which are yet to come
- Reformulate the HEP examples

bluehood committed:
Merge pull request #68 from dpiparo/user_guide

Various improvements, among which:

* comment out description of features that are not yet present
* reformulate HEP examples

eguiraud committed:
fix typos and such

eguiraud committed:
other minor improvements

eguiraud committed:
improvements to tables

eguiraud committed:
Get -> Take

eguiraud committed:
delayed actions -> lazy actions

bluehood commented:
I believe the user guide is ready to be included in master. Further improvements and incorporation of new feedback can be done directly on master.
